# Practice Notebook — Session 6 (Optimization & Gradient Descent)
Problems with answers for optimization basics and gradient methods.

Generated: 2025-09-05T16:30:05.298840 UTC

---

## Problem 1 — Analytic minimizer
Find the minimizer of f(x) = (x-3)^2 and the minimum value.

In [1]:
# Problem 1 — analytic
# f(x) = (x-3)^2
x_star = 3.0
f_min = 0.0
print('Minimizer x* =', x_star)
print('Minimum value f(x*) =', f_min)

Minimizer x* = 3.0
Minimum value f(x*) = 0.0


**Answer (computed):** x* = `3.0`, f(x*) = `0.0`

## Problem 2 — Gradient Descent (1D)
Run gradient descent on f(x) = (x-3)^2 starting from w0=0, lr=0.1 for 10 epochs and show the parameter trajectory.

In [2]:
# Problem 2 — gradient descent
w = 0.0
lr = 0.1
history = []
for epoch in range(10):
    grad = 2*(w-3)
    w = w - lr * grad
    history.append((epoch+1, w, (w-3)**2))
print('Epoch, w, f(w)')
for rec in history:
    print(rec)

Epoch, w, f(w)
(1, 0.6000000000000001, 5.76)
(2, 1.08, 3.6864)
(3, 1.464, 2.359296)
(4, 1.7711999999999999, 1.5099494400000002)
(5, 2.01696, 0.9663676415999999)
(6, 2.213568, 0.6184752906240001)
(7, 2.3708544, 0.39582418599936026)
(8, 2.49668352, 0.25332747903959046)
(9, 2.597346816, 0.1621295865853379)
(10, 2.6778774528, 0.10376293541461637)


**Answer (computed sample):** first few iterations shown in code output; final w ≈ `2.6778774528`

## Problem 3 — Stochastic Gradient Descent (SGD)
Perform SGD for linear regression y=2x (perfect linear) with lr=0.01 for 5 epochs. Show final parameters.

In [3]:
# Problem 3 — SGD example
import numpy as np
X = np.array([[1],[2],[3],[4]], dtype=float)
y = np.array([2,4,6,8], dtype=float)
W, b = 0.0, 0.0
lr = 0.01
for epoch in range(5):
    for i in range(len(X)):
        xi = X[i]
        yi = y[i]
        y_pred = W*xi + b
        error = y_pred - yi
        dW = 2 * xi * error
        db = 2 * error
        W -= lr * dW
        b -= lr * db
print('Final W, b after SGD:', W, b)

Final W, b after SGD: [1.77443669] [0.59037272]


**Answer (computed):** final W ≈ `[1.77443669]`, b ≈ `[0.59037272]` (after 5 epochs in this example)

## Problem 4 — Convexity check
Check convexity of f(x) = x^2 + 2x + 3 using second derivative.

In [4]:
# Problem 4 — convexity
# f''(x) = 2 > 0 -> convex
print('Second derivative = 2 -> convex')

Second derivative = 2 -> convex


**Answer:** convexity = `True`